### PR TITLE
Idempotently delete volumes during disk delete

### LINKED ISSUE #3305: Saga action error returned when dealing with multiple concurrent disk deletions
This seems to be a new behavior after the [fix for saga undo action](https://github.com/oxidecomputer/steno/pulls/138).

I have a Terraform plan that destroys 3 instances and their disks (2 disks per instance). The 6 concurrent disk deletions result in 500 errors even though the disks all ended up being deleted successfully:
```
$ terraform destroy
...
Plan: 0 to add, 0 to change, 9 to destroy.

Do you really want to destroy all resources?
  Terraform will destroy all your managed infrastructure, as shown above.
  There is no undo. Only 'yes' will be accepted to confirm.

  Enter a value: yes

oxide_instance.app_instance[2]: Destroying... [id=dabd83bd-51fc-40f7-a444-1e0859098eba]
oxide_instance.app_instance[0]: Destroying... [id=01fddbd2-e5e2-463f-bbdc-060dec72bb2e]
oxide_instance.app_instance[1]: Destroying... [id=7d61b2ed-b0a8-4f28-851e-80454b62968c]
oxide_instance.app_instance[0]: Destruction complete after 3s
oxide_instance.app_instance[2]: Destruction complete after 3s
oxide_instance.app_instance[1]: Destruction complete after 8s
oxide_disk.boot[1]: Destroying... [id=30e364d0-455f-4496-ad3a-7e545bae7f00]
oxide_disk.data[0]: Destroying... [id=f1ccfbfe-f809-42eb-82d2-982bd931c840]
oxide_disk.data[1]: Destroying... [id=b4cdef5c-326c-4ab8-811b-544173589375]
oxide_disk.boot[0]: Destroying... [id=5cdc3869-8844-4037-b3e8-ef8b4d4e0e17]
oxide_disk.boot[2]: Destroying... [id=fc03c979-39af-44e4-9c23-d2fb95d295ba]
oxide_disk.data[2]: Destroying... [id=cb39dd1b-8acf-49f5-9b5b-638ada87ca32]
╷
│ Error: Unable to delete disk:
│ 
│ API error: HTTP 500 (https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840) BODY -> {
│   "request_id": "bd6e9f4f-37b4-42b0-892d-027c80732000",
│   "error_code": "Internal",
│   "message": "Internal Server Error"
│ }
╵
╷
│ Error: Unable to delete disk:
│ 
│ API error: HTTP 500 (https://recovery.sys.rack2.eng.oxide.computer/v1/disks/cb39dd1b-8acf-49f5-9b5b-638ada87ca32) BODY -> {
│   "request_id": "192f48a6-403f-4a0d-a880-0717e440c202",
│   "error_code": "Internal",
│   "message": "Internal Server Error"
│ }
╵
╷
│ Error: Unable to delete disk:
│ 
│ API error: HTTP 500 (https://recovery.sys.rack2.eng.oxide.computer/v1/disks/fc03c979-39af-44e4-9c23-d2fb95d295ba) BODY -> {
│   "request_id": "780be980-d2bd-409f-8e01-8930ddf6f5a6",
│   "error_code": "Internal",
│   "message": "Internal Server Error"
│ }
╵
╷
│ Error: Unable to delete disk:
│ 
│ API error: HTTP 500 (https://recovery.sys.rack2.eng.oxide.computer/v1/disks/b4cdef5c-326c-4ab8-811b-544173589375) BODY -> {
│   "request_id": "bf987b99-22b6-4100-91ba-24717ed82145",
│   "error_code": "Internal",
│   "message": "Internal Server Error"
│ }
╵
╷
│ Error: Unable to delete disk:
│ 
│ API error: HTTP 500 (https://recovery.sys.rack2.eng.oxide.computer/v1/disks/30e364d0-455f-4496-ad3a-7e545bae7f00) BODY -> {
│   "request_id": "2cfa2fc2-efdc-44bb-818e-428f4a63aee5",
│   "error_code": "Internal",
│   "message": "Internal Server Error"
│ }
╵
╷
│ Error: Unable to delete disk:
│ 
│ API error: HTTP 500 (https://recovery.sys.rack2.eng.oxide.computer/v1/disks/5cdc3869-8844-4037-b3e8-ef8b4d4e0e17) BODY -> {
│   "request_id": "935d0d25-ac49-4c41-9bd2-2c4f91cfb587",
│   "error_code": "Internal",
│   "message": "Internal Server Error"
│ }
```
Here are the detailed log messages for one of the disk deletion requests:
```
root@oxz_nexus:~# grep bd6e9f4f-37b4-42b0-892d-027c80732000 /var/svc/log/oxide-nexus\:default.log | looker       
15:32:08.125Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): roles
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    roles = RoleSet { roles: {} }
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.125Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): authorize result
    action = Query
    actor = Some(Actor::SiloUser { silo_user_id: 001de000-05e4-4000-8000-000000004007, silo_id: 001de000-5110-4000-8000-000000000000, .. })
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    resource = Database
    result = Ok(())
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.126Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): roles
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    roles = RoleSet { roles: {} }
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.126Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): authorize result
    action = Query
    actor = Some(Actor::SiloUser { silo_user_id: 001de000-05e4-4000-8000-000000004007, silo_id: 001de000-5110-4000-8000-000000000000, .. })
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    resource = Database
    result = Ok(())
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.127Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): roles
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    roles = RoleSet { roles: {} }
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.128Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): authorize result
    action = Query
    actor = Some(Actor::SiloUser { silo_user_id: 001de000-05e4-4000-8000-000000004007, silo_id: 001de000-5110-4000-8000-000000000000, .. })
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    resource = Database
    result = Ok(())
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.129Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): roles
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    roles = RoleSet { roles: {} }
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.129Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): authorize result
    action = Query
    actor = Some(Actor::SiloUser { silo_user_id: 001de000-05e4-4000-8000-000000004007, silo_id: 001de000-5110-4000-8000-000000000000, .. })
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    resource = Database
    result = Ok(())
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.131Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): roles
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    roles = RoleSet { roles: {} }
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.132Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): authorize result
    action = Query
    actor = Some(Actor::SiloUser { silo_user_id: 001de000-05e4-4000-8000-000000004007, silo_id: 001de000-5110-4000-8000-000000000000, .. })
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    resource = Database
    result = Ok(())
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.134Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): roles
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    roles = RoleSet { roles: {} }
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.134Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): authorize result
    action = Query
    actor = Some(Actor::SiloUser { silo_user_id: 001de000-05e4-4000-8000-000000004007, silo_id: 001de000-5110-4000-8000-000000000000, .. })
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    resource = Database
    result = Ok(())
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.136Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): roles
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    roles = RoleSet { roles: {(Fleet, 001de000-1334-4000-8000-000000000000, "admin"), (Silo, 001de000-5110-4000-8000-000000000000, "admin")} }
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.140Z DEBG 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): authorize result
    action = Delete
    actor = Some(Actor::SiloUser { silo_user_id: 001de000-05e4-4000-8000-000000004007, silo_id: 001de000-5110-4000-8000-000000000000, .. })
    actor_id = 001de000-05e4-4000-8000-000000004007
    authenticated = true
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    resource = Disk { parent: Project { parent: Silo { parent: Fleet, key: 001de000-5110-4000-8000-000000000000, lookup_type: ById(001de000-5110-4000-8000-000000000000) }, key: 1ec35f61-0fb3-4fe7-a460-f930eff468f6, lookup_type: ById(1ec35f61-0fb3-4fe7-a460-f930eff468f6) }, key: f1ccfbfe-f809-42eb-82d2-982bd931c840, lookup_type: ById(f1ccfbfe-f809-42eb-82d2-982bd931c840) }
    result = Ok(())
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
15:32:08.259Z INFO 612405c5-68bb-41a8-905c-e729dfd7514d (dropshot_external): request completed
    error_message_external = Internal Server Error
    error_message_internal = saga ACTION error at node "no_result2": saga ACTION error at node "no_result_3": Internal Server Error
    local_addr = 172.30.1.5:443
    method = DELETE
    remote_addr = 172.20.17.42:57144
    req_id = bd6e9f4f-37b4-42b0-892d-027c80732000
    response_code = 500
    uri = https://recovery.sys.rack2.eng.oxide.computer/v1/disks/f1ccfbfe-f809-42eb-82d2-982bd931c840
```

### PR DESCRIPTION
**Pull request by jmpesp**
Previous to this commit, deleting a volume during the disk delete saga was done by starting the execution of the `volume_delete` saga, but this is not idempotent. Embed a `volume_delete` subsaga instead.

There's two other sagas where `Nexus::volume_delete` is called: during the disk creation saga and the snapshot creation saga, both when unwinding the saga. Unfortunately, embedding a subsaga cannot be done as part of the *rollback* of a saga, because there's no way to create an undo node that itself is a subsaga. This commit retains the previous behaviour during unwind: kick off the `volume_delete` saga. If there's a failure, then the saga will be parked as "Stuck" anyway.

Fixes #3305

### CODE REVIEW

File: nexus/src/app/sagas/disk_delete.rs [hunk 1]
```diff
@@ -54,12 +55,43 @@ impl NexusSaga for SagaDiskDelete {
...
+                    "params_for_volume_delete_subsaga".to_string(),
+                    e,
+                )
+            })?,
+        ));
+
+        builder.append(Node::subsaga(
+            "volume_delete_subsaga_no_result",
+            subsaga_dag,
+            "params_for_volume_delete_subsaga",
```
Comment (davepacheco) [inline nexus/src/app/sagas/disk_delete.rs:92]: nit: I think this was copied/pasted but should probably say something else

**Review by davepacheco — APPROVED** [review]
This change itself looks fine (aside from one nit below).

I'm not clear on why the undo case isn't parallel... but I think that's probably outside the scope of this PR.

File: nexus/src/app/sagas/disk_delete.rs (hunk 1 above)
Comment (jmpesp) [reply nexus/src/app/sagas/disk_delete.rs:92]: Sorry, I'm not against this but I don't understand the suggestion - which part? Is there a correctness issue? What other name(s) do you have in mind?

**Comment by jmpesp** [comment]
> I'm not clear on why the undo case isn't parallel... but I think that's probably outside the scope of this PR.

Basically, (as far as I know) there's no way right now for an undo node in the DAG to point to a subsaga of action and undo nodes.

**Comment by davepacheco** [comment]
> > I'm not clear on why the undo case isn't parallel... but I think that's probably outside the scope of this PR.
> 
> Basically, (as far as I know) there's no way right now for an undo node in the DAG to point to a subsaga of action and undo nodes.

Yeah, that is right, but I'm not clear on the use case for that.  Usually you break up the saga into actions and each has an undo action.  If you use a subsaga to do a bunch of stuff, then I'd expect the undo actions for the subsaga to undo it.  It seems surprising to have a node whose action is simple but whose undo action is complicated enough to want to be a whole subsaga.  To be clear I haven't looked closely at this code and I'm not saying that's wrong!

File: nexus/src/app/sagas/disk_delete.rs (hunk 1 above)
Comment (davepacheco) [reply nexus/src/app/sagas/disk_delete.rs:92]: I misread this.  This is perfect as-is.  Sorry!

**Pull request merged**